source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,659 @@
+
+///
+unittest
+{
+  import std.format;
+  import std.traits;
+
+  interface GrandTour
+  {
+    pure int foo() immutable;
+    @nogc @trusted nothrow ref int foo(out real, return ref int, lazy int) const;
+    @safe shared scope void bar(scope Object);
+  }
+
+  class Mock(Interface) : Interface
+  {
+    static foreach (member; __traits(allMembers, Interface)) {
+        static foreach (fun; __traits(getOverloads, Interface, member)) {
+          mixin({
+              enum Model = refract!(fun, "fun");
+              if (is(ReturnType!fun == void)) {
+                return Model.setBody("{}").mixture;
+              } else if (Model.attributes & FunctionAttribute.ref_) {
+                return Model.setBody(q{{
+                    static %s rv;
+                    return rv;
+                  }}.format(Model.returnType)).mixture;
+              } else {
+                return Model.setBody(q{{
+                    return %s.init;
+                  }}.format(Model.returnType)).mixture;
+              }
+            }());
+        }
+    }
+  }
+
+  GrandTour mock = new Mock!GrandTour;
+  real x;
+  int i, l;
+  mock.foo(x, i, l++) = 1;
+  assert(mock.foo(x, i, l++) == 1);
+  assert(l == 0);
+}
+
+private enum isAggregate(T...) =
+  is(T[0] == struct) || is(T[0] == union) || is(T[0] == class)
+  || is(T[0] == interface);
+
+/**
+    A struct capturing all the properties of a function.
+*/
+
+private mixin template replaceAttribute(string Name)
+{
+  alias Struct = typeof(this);
+  mixin(
+    "Struct copy = {",
+    {
+      string[] mixture;
+      foreach (member; __traits(allMembers, Struct)) {
+        if (__traits(getOverloads, Struct, member).length == 0) {
+          //pragma(msg, member);
+          mixture ~= member ~ ":" ~ (member == Name ? "value" : member);
+        }
+      }
+      return mixture.join(",\n");
+    }(),
+    "};"
+  );
+}
+
+
+immutable struct Function
+{
+
+  /**
+     A string that evaluates to a function symbol.
+   */
+
+  string localSymbol;
+
+  /**
+     Function name. Initial value: `__traits(identifier, fun)`.
+   */
+
+  string name;
+
+  /**
+     Return a new Function object with the name attribute set to value.
+  */
+
+  Function setName(string value)
+  {
+    mixin replaceAttribute!"name";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    pure @nogc int answer();
+    mixin(refract!(answer, "answer").setName("ultimateAnswer").mixture);
+    static assert(
+      __traits(getAttributes, ultimateAnswer) ==
+      __traits(getAttributes, answer));
+  }
+
+  /**
+     Index of function in a set of overloads. Valid only if Function represents
+     a function (not a function pointer), and was refracted from a module.
+   */
+
+  uint index;
+
+  /**
+     Return a new Function object with the index attribute set to value.
+  */
+
+  Function setIndex(uint value)
+  {
+    mixin replaceAttribute!"index";
+    return copy;
+  }
+
+  /**
+     Return a string representing the parameter.
+   */
```

Comment:
This commend has not target? Is it missing a member?

Reply:
> This comment has not target? Is it missing a member?

Removed.